src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/CollectionsMarshalTests.cs

```diff
@@ -299,6 +299,200 @@ public void GetValueRefOrNullRefLinkBreaksOnResize()
             Assert.Equal(50, dict.Count);
         }
 
+        [Fact]
+        public void GetValueRefOrAddDefaultValueType()
+        {
+            // This test is the same as the one for GetValueRefOrNullRef, but it uses
+            // GetValueRefOrAddDefault instead, and also checks for incorrect additions.
+            // The two APIs should behave the same when values are already existing.
+            var dict = new Dictionary<int, Struct>
+            {
+                {  1, default },
+                {  2, default }
+            };
+
+            Assert.Equal(2, dict.Count);
+
+            Assert.Equal(0, dict[1].Value);
+            Assert.Equal(0, dict[1].Property);
+
+            var itemVal = dict[1];
+            itemVal.Value = 1;
+            itemVal.Property = 2;
+
+            // Does not change values in dictionary
+            Assert.Equal(0, dict[1].Value);
+            Assert.Equal(0, dict[1].Property);
+
+            CollectionsMarshal.GetValueRefOrAddDefault(dict, 1, out bool exists).Value = 3;
+
+            Assert.True(exists);
+            Assert.Equal(2, dict.Count);
+
+            CollectionsMarshal.GetValueRefOrAddDefault(dict, 1, out exists).Property = 4;
+
+            Assert.True(exists);
+            Assert.Equal(2, dict.Count);
+            Assert.Equal(3, dict[1].Value);
+            Assert.Equal(4, dict[1].Property);
+
+            ref var itemRef = ref CollectionsMarshal.GetValueRefOrAddDefault(dict, 2, out exists);
+
+            Assert.True(exists);
+            Assert.Equal(2, dict.Count);
+            Assert.Equal(0, itemRef.Value);
+            Assert.Equal(0, itemRef.Property);
+
+            itemRef.Value = 5;
+            itemRef.Property = 6;
+
+            Assert.Equal(5, itemRef.Value);
+            Assert.Equal(6, itemRef.Property);
+            Assert.Equal(dict[2].Value, itemRef.Value);
+            Assert.Equal(dict[2].Property, itemRef.Property);
+
+            itemRef = new() { Value = 7, Property = 8 };
+
+            Assert.Equal(7, itemRef.Value);
+            Assert.Equal(8, itemRef.Property);
+            Assert.Equal(dict[2].Value, itemRef.Value);
+            Assert.Equal(dict[2].Property, itemRef.Property);
+
+            // Check for correct additions
+
+            ref var entry3Ref = ref CollectionsMarshal.GetValueRefOrAddDefault(dict, 3, out exists);
+
+            Assert.False(exists);
+            Assert.Equal(3, dict.Count);
+            Assert.False(Unsafe.IsNullRef(ref entry3Ref));
```

Comment:
Shouldn't there also be an assertion here validating that `entry3Ref` equals `default(Struct)`?

Reply:
Right, yeah it was present in the ref type case below but not here. Fixed in 21ab729485561a6c292a90ebc5a1c28182d3f64a, thanks! 😄